data_processing/process_data.py

```diff
@@ -17,11 +17,10 @@ def create_interaction_matrix(directory_path, num_users, num_movies, files):
                 user_id = int(user_id)
                 rating = int(rating)
                 
-                if rating in (4, 5):
-                    if user_id not in user_mapper:
+                if rating in (4, 5) and user_id not in user_mapper: # used AND operator instead nested IF
```

Comment:
I think by combining these if statements you've made the logic incorrect. You want to store the rating in the `interaction_matrix1 even if the user_id is already in the `user_mapper`, right?

So if you see user 1400, and you map it to user 235, you still want to save a second rating for this user if you see it again, right?

Reply:
What I was suggesting before was "returning early". Here's an example with fake function names:

```
def calculate_data(db):
  if db.is_open():
    conn = db.get_conn()
    cursor = conn.get_cursor()
    data = cursor.fetch_data()
    if len(data) > 1:
      update_data(data)
      data.process()
```

Notice that the entire function is indented under an `if` statement. Same thing in the inner if, all of the code is nested underneath it. If we reverse the if condition, we can simply return early and skip the rest of the function. This means that the rest of the function doesn't have to be indented, but more importantly it removes the cognitive overhead when reading the code. Instead of being N levels deep and having to remember "Okay, if I'm here then this and this and this must be true", you can simply reason "okay if there's no db connection we return early" and not have to think about it further.

Here's the same example with "early returns":

```
def calculate_data(db):
  if not db.is_open():
   return

  conn = db.get_conn()
  cursor = conn.get_cursor()
  data = cursor.fetch_data()

  if not len(data) > 1:
    return
   
  update_data(data)
  data.process()
```

Reply:
In your case, since you're in a for loop, it would be early `continue` instead of early `return`.

Reply:
Something like this?

```python
            for line in file:
                user_id, rating, _ = line.strip().split(',')
                user_id = int(user_id)
                rating = int(rating)
                
                if rating < 4:
                    continue

                if user_id not in user_mapper:
                    user_mapper[user_id] = current_user_index
                    current_user_index += 1
                    user_idx = user_mapper[user_id]
                    interaction_matrix[user_idx, movie_idx] = rating

                return interaction_matrix


```

Reply:
Esentially yes, though the logic in the second if statement is wrong as you wrote it here, because you need to get the `user_idx` and store `rating` in the `interaction_matrix` whether or not you had to remap the id.